Update walkthrough to remove MDC (#27)
aerogear/mobile-walkthrough: walkthroughs/1-ionic-showcase/walkthrough.adoc

> = Running the mobile showcase application with {mobile-services-name}

Reviewer: I am thinking maybe we should call it "GraphQL showcase application" rather than "mobile". Also I think we should remove `{mobile-services-name}` at the end.

@pwright wdyt?

Author: Awesome idea!

This Solution Pattern demonstrates running the GraphQL PWA Showcase Application using Red Hat Mobile Developer Services.

> Solution Pattern will show you how to:

Reviewer: Not sure if this change is required. Also `Pattern` should be in lower case.

> * Build and run the link:{showcase-app-link}[Ionic Mobile Showcase app].

Reviewer: may be we just call it `GraphQL client app`?

Author: It has push integration so showcase still stands :) plus is nice to know the technology - Ionic as it might be appealing to some people

* Deploy the data sync server of the GraphQL Showcase app.

> * Protect the showcase GraphQL app and the backend service using the Red Hat SSO.

Reviewer 2: ```suggestion
* Protect the showcase GraphQL app and the backend service using Red Hat SSO.
```

> * Protect the Showcase GraphQL app and the backend service using the Red Hat SSO.

Reviewer: `Showcase` -> `showcase`

> It will also introduce you to all features offered by the AeroGear Data Sync framework.

Reviewer 2: ```suggestion
It will also introduce you to all of the features offered by the AeroGear Data Sync framework.
```

> It will also introduce you to all features offered by the DataSync framework.

Reviewer: is it right to use `DataSync` here? What do we use in the docs?

* link:{openshift-host}/console/project/{walkthrough-namespace}/overview[OpenShift Project, window="_blank"]
****

[type=walkthroughResource]
.{data-sync-name} Documentation
****

> * link:https://github.com/aerogear/voyager-server/tree/master/doc[Getting Started with {data-sync-name}, window="_blank"]

Reviewer: @pwright are we going to have a link of the data sync docs on customer portal, or it is ok if we joint point to the upstream docs for now?

Author: Upstream docs will not exist. I think we agreed with Paul that docs in voyager server will feed directly into access.redhat.com. What it means that we probably might be better with previos link.

@darahayes to confirm

Verify that you followed each step in the procedure above.  If you are still having issues, contact your administrator.
****

[time=10]
== Access the showcase app in a browser

> The showcase app demonstrates the key capabilities provided by AeroGear Data Sync framework

Reviewer 2: ```suggestion
The showcase app demonstrates the key capabilities provided by the AeroGear Data Sync framework
```

Verify that you followed each step in the procedure above.  If you are still having issues, contact your administrator.
****

[time=10]

> == Access Showcase PWA App in a browser

Reviewer: ```suggestion
== Access the showcase app in a browser
```

> The showcase app demonstrates the key capabilities provided by {mobile-services-name}.

Reviewer: probably can replace `{mobile-services-name}` with `AeroGear data sync framework`.
```suggestion
The showcase app demonstrates the key capabilities provided by AeroGear Data Sync framework
```

It can run either as a hybrid mobile application, or a Progressive Web App (PWA).

> In this task, we will access PWA application that is hosted by server side container that you can run in your browser.

Reviewer: I think we can just remove this line. It's not really necessary.
```suggestion
```

> . Go to link:{route-ionic-showcase-server-host}[GraphQL playground, window="_blank"]

Reviewer: is it actually the playground?
```suggestion
. Go to link:{route-ionic-showcase-server-host}[Showcase app, window="_blank"]
```

****

[time=15]
== Protect the showcase app using the Red Hat SSO

> In this task, we will protect both the frontend and the backend of the {data-sync-showcase-app} with the {customer-sso-name}.

Author: I'm not big fan of the `protect` word as it is not universally translating to `add authz`. Not big issue

****

[time=15]
== Protect the showcase app using the Red Hat SSO

In this task, we will protect both the frontend and the backend of the {data-sync-showcase-app} with the {customer-sso-name}.

=== Configure and protect the frontend app

. Make sure you have access to a realm in the {customer-sso-name}. If you don't, you can ask your customer-admin user to create one for you. 
. Login to the realm in the {customer-sso-name}.

> . Create a new client for the frondend:

Author: ```suggestion
. Create a new client for the frontend app:
```

.. Choose *Clients* under *Configure* in the menu.
.. Click on the *Create* button on the top-right corner to start creating a new client.
.. Enter `showcase-app-client` for Client ID.  Do not change the rest of the fields.
.. Save the client. You should see the *Settings* page of the client.
.. You should see the *Settings* page of the client.
.. Change the `Valid Redirect URIs` to '{route-ionic-showcase-server-host}*'
.. Change the `Web Origins` to '*'
.. Click on the *Installation* tab, and select `Keycloak OIDC JSON` format. Copy the content or use the `Download` button to save the configuration file.

Next we need to update the configuration of the frontend app to secure it:

. Navigate to the link:{openshift-host}/console/project/{walkthrough-namespace}/browse/config-maps[Config Maps page of the OpenShift console].
. Select the config map that is called `webapp-config`, and edit it.
. Add a new `auth` section by pasting the content that was copied in the previous step.
.. You should then replace `auth-server-url` to `url` and `resource` to `clientId`.
. The content of the config map should look as follows:

> 

Author: ```suggestion
+
```

---
    window.showcaseConfig = {
      "backend": {
        "serverUrl": "/graphql",
        "wsServerUrl": ((window.location.protocol === "https:") ? "wss://" : "ws://") + window.location.hostname + "/graphql"
      }
      "auth": {
        "realm": "your realm",
        "url": "https://your-server/auth",
        "clientId": "your-client"
      }
    };
---

> 

Author: ```suggestion
+
```

****

[time=15]

> == Protect the GraphQL showcase app using the Red Hat SSO

Reviewer 2: ```suggestion
== Protect the GraphQL showcase app using Red Hat SSO
```

> In this task, you use the Red Hat SSO to protect both the client (PWA) and the data sync server.

Reviewer 2: ```suggestion
In this task, you use Red Hat SSO to protect both the client (PWA) and the data sync server.
```

=== Connect SSO to the client app

. Make sure you have access to a realm in the user-facing SSO instance. If you don't have access, you can ask your customer-admin user to create one for you. 

> . In new realm we going to create two new clients - one for server side and one for the client side application.

Reviewer 2: This isn't really a step in the procedure, it's more of a description of what we're going to do. I think you could put this up under the `Connect SSO to the client app` heading but before the bullet points. Also it's got some grammar that native speakers will find strange.

I would maybe reword it to the following.

`In the following steps, we will create a new keycloak realm. In the new realm we will create two keycloak clients. One for the server side, and one for the client side application'

. Open URL obtained from adminstrator
. Login to keycloak (you will be asked to provide additional details like email and name)

> . When completing login process you should see the main keycloak realm page

Reviewer 2: ```suggestion
. You should see the main Keycloak realm page when logged in
```

> . In the right menu choose *Clients* under *Configure* in the menu.

Reviewer 2: Is this in the on the right side? Is it not on the left side?

Also you're saying 'in the menu choose x in the menu'

Maybe it could be something like 

```suggestion
. In the menu on the right side, choose *Configure* > *Clients*.
```

> .. Click on the *Create* button on the top-right corner to start creating a new client.

Reviewer 2: ```suggestion
.. Click on the *Create* button on the top-right corner to create a new client.
```

> .. Enter `showcase-app-client` for Client ID.  Do not change the rest of the fields.

Reviewer 2: ```suggestion
.. Enter `showcase-app-client` for Client ID. Do not change any other fields.
```

.. Save the client. You should see the *Settings* page of the client.

> .. You should see the *Settings* page of the client.

Reviewer 2: Duplicate of the line above

> .. Change the `Valid Redirect URIs` to '{route-ionic-showcase-server-host}*'

Reviewer 2: ```suggestion
.. Change `Valid Redirect URIs` to '{route-ionic-showcase-server-host}*'
```

> .. Change the `Web Origins` to '*'

Reviewer 2: ```suggestion
.. Change `Web Origins` to '*'
```

> .. Click on the *Installation* tab, and select `Keycloak OIDC JSON` format. Use the `Download` button to save the configuration file to a directory on the computer.

Reviewer 2: ```suggestion
.. Click on the *Installation* tab, and select `Keycloak OIDC JSON` format. Use the `Download` button to save the configuration file to a directory on your computer.
```

. Navigate to the project config-maps section {openshift-host}/console/project/{walkthrough-namespace}/browse/config-maps
. Select `webapp-config`
. Add new `auth` section in config by utilizing the config that was copied in previous step.
. In configuration replace `auth-server-url` to `url` and `resource` to `clientId`

> . Configuration should look as follows

Reviewer 2: ```suggestion
. The configuration should look similar to the one below.
```

---
    window.showcaseConfig = {
      "backend": {
        "serverUrl": "/graphql",
        "wsServerUrl": ((window.location.protocol === "https:") ? "wss://" : "ws://") + window.location.hostname + "/graphql"
      }
      "auth": {

>         "realm": "your realm",

Reviewer 2: ```suggestion
        "realm": "<your realm>",
```

        "url": "https://your-server/auth",

>         "clientId": "your-client"

Reviewer 2: ```suggestion
        "clientId": "<your-client>"
```

      }
    };
---

> . Save and wait for server to redeploy

Reviewer 2: ```suggestion
. Save and wait for the server to redeploy
```

> === Add Identity Management service to the client app (PWA) and create a realm for users

Reviewer: actually if you read the title here, I think this section is just about creating the realm. The next section is actually about creating the clients. So we probably should move some of the content in this section to the next one, or create a different section to cover how to get KC config for the client app.

> . As your admin to create new realm for your user and obtain url to be able to login to the keycloak instance

Reviewer: ```suggestion
. Make sure you have access to a realm in the user-facing SSO instance. If you don't have access, you can ask your customer-admin user to create one for you. 
```

> . In new realm we going to create two new clients - one for server side and one for the client side application.

Reviewer: is this line part of the instruction? if not we can remove this.

Reviewer: Also the steps from here doesn't really belong in this section: it's about getting the KC client config for the client app.

Author: Got it cleared based on my understanding but feel free to rebuild it

> . Open URL obtained from adminstrator

Author: This is the most problematic part. Generally eval users will not be able to finish this walkthough without intervence of the admin user. I think @pwright has really good proposal to address this problem and this use case can be used as good example when this is needed. 

Fix for that (in integrately) will be very simple - we need to get some extra realm with users that will be matching the users created on the cluster realm (the same password). This will enable developers to explore user facing keycloak without intervention of the cluster or customer admins.

CC @wei-lee

Reviewer 3: @wtrocki  are you proposing a change to the integreatly installer? If so, we need that to happen before this walkthrough works, right?

Author: @pwright I'm just flagging problem in walkthough - technically people might be able to get access to admin and create realm thenselves etc. so it might not be that kind of issue but it is possible area to improve. I just suggested possible way but maybe there is another way to do it.

> . Login to keycloak (you will be asked to provide additional details like email and name)

Reviewer: we can probably merge the previous step and this one together. Something like:

```
. Get the URL of the keycloak instance and login with the credential provided by the customer-admin user.
```

. When completing login process you should see the main keycloak realm page
. In the right menu choose *Clients* under *Configure* in the menu.
.. Click on the *Create* button on the top-right corner to start creating a new client.
.. Enter `showcase-app-client` for Client ID.  Do not change the rest of the fields.
.. Save the client. You should see the *Settings* page of the client.
.. You should see the *Settings* page of the client.
.. Change the `Valid Redirect URIs` to '{route-ionic-showcase-server-host}*'
.. Change the `Web Origins` to '*'
.. Click on the *Installation* tab, and select `Keycloak OIDC JSON` format. Use the `Download` button to save the configuration file to a directory on the computer.

. Navigate to the project config-maps section {openshift-host}/console/project/{walkthrough-namespace}/browse/config-maps
. Select `webapp-config`
. Add new `auth` section in config by utilizing the config that was copied in previous step.

> . In configuration replace `auth-server-url` to `url` and `resource` to `clientId`

Author: This can be done in showcase but it is more valuable to show users that this needs to be done for their apps rather than hiding this fact in the showcase implementation.

Reviewer: is there any particular reason that we ask developer to modify the downloaded config for the showcase app?

Author: it is not for showcase app itself. It is that SSO adapter for js require it (even in their docs).

Author: https://www.keycloak.org/docs/latest/securing_apps/index.html#_javascript_adapter :)

The format that SSO provides is old because I think we use some old version there.

. Configuration should look as follows

---
    window.showcaseConfig = {
      "backend": {
        "serverUrl": "/graphql",
        "wsServerUrl": ((window.location.protocol === "https:") ? "wss://" : "ws://") + window.location.hostname + "/graphql"
      }
      "auth": {

>         "realm": "your realm",

Author: We do not know realm name. Walkthough can assume that this can be master realm and then we can fill those fields here automatically so developers will be able to copy them.
We will have keycloak url, realm and clientId (as we will ask users to create it with a certain name)

        "url": "https://your-server/auth",
        "clientId": "your-client"
      }
    };
---

. Save and wait for server to redeploy

> === Configure Identity Management service for the showcase client and server applications

Reviewer: we might want to change the title here as we only cover the server part.
```suggestion
=== Configure Identity Management service for the showcase client applications
```
